crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+                }
+                crate::ComponentExportKind::Component(idx) => {
+                    insert_export(
+                        export.name,
+                        ComponentEntityType::Component(self.component_at(*idx, offset)?.clone()),
+                        &mut inst_exports,
+                        offset,
+                    )?;
+                }
+                crate::ComponentExportKind::Instance(idx) => {
+                    insert_export(
+                        export.name,
+                        ComponentEntityType::Instance(
+                            self.component_instance_at(*idx, offset)?.clone(),
+                        ),
+                        &mut inst_exports,
+                        offset,
+                    )?;
+                }
+                crate::ComponentExportKind::Function(idx) => {
+                    insert_export(
+                        export.name,
+                        ComponentEntityType::Func(
+                            self.component_function_at(*idx, offset)?.clone(),
+                        ),
+                        &mut inst_exports,
+                        offset,
+                    )?;
+                }
+                crate::ComponentExportKind::Value(idx) => {
+                    insert_export(
+                        export.name,
+                        ComponentEntityType::Value(self.value_at(*idx, offset)?.clone()),
+                        &mut inst_exports,
+                        offset,
+                    )?;
+                }
+                crate::ComponentExportKind::Exports(exports) => {
+                    insert_export(
+                        export.name,
+                        ComponentEntityType::Instance(
+                            self.instantiate_exports(exports.as_ref(), offset)?,
+                        ),
+                        &mut inst_exports,
+                        offset,
+                    )?;
+                }
+            }
+        }
+
+        Ok(InstanceType {
+            exports: Rc::new(inst_exports),
+        })
+    }
+
+    fn alias_instance_export(
+        &mut self,
+        kind: crate::AliasKind,
+        idx: u32,
+        name: &str,
+        offset: usize,
+    ) -> Result<()> {
+        macro_rules! push_module_export {
+            ($expected:path, $collection:ident, $ty:literal) => {{
+                match self.module_instance_export(idx, name, offset)?.clone() {
+                    $expected(ty) => {
+                        self.$collection.push(ty);
+                        Ok(())
+                    }
+                    _ => {
+                        return Err(BinaryReaderError::new(
+                            format!("export `{}` for instance {} is not a {}", name, idx, $ty),
+                            offset,
+                        ))
+                    }
+                }
+            }};
+        }
+
+        macro_rules! push_component_export {
```

Comment:
This feels as good a place as any to comment on this, but from the spec level I'm not really convinced at this point that separate module and component types are worth it. So far everywhere in the validator seems to duplicate almost the same logic for component instances vs module instances. It seems to me like there should be one "module type" and one "instance type" which everything is lumped into. I may need to bring this up with the spec though? It just feels like unnecessary hoop dancing to have these internal distinctions in so many places.

Reply:
Yeah, let's bring that up with the spec. I believe one of the other spec authors proposed the dual types, but the specific reasons for it are escaping me right now.

Reply:
Resolving this spec issue for now.